composer/core/state.py

```diff
@@ -501,6 +525,19 @@ def algorithms(self):
     def algorithms(self, algorithms: Sequence[Algorithm]):
         self._algorithms[:] = algorithms
 
+    def _dataset_of(self, obj: Optional[Union[Iterable, Evaluator, DataSpec, DataLoader]]) -> Optional[Dataset]:
+        from composer.core.evaluator import Evaluator
+        if obj is None:
+            return None
+        if isinstance(obj, Evaluator):
+            obj = obj.dataloader
```

Comment:
Should this be `obj.dataloader.dataset` ?

Reply:
"obj" (the "dataloader") may be an Evaluator. if it is, it has a "dataloader" field

that "dataloader" may actually be a DataSpec (DataLoader + on-device transform). if it is, it has a "dataloader" field

that result will be the actual DataLoader (return obj.dataset), unless the user has passed some other kind of Iterable in which case it doesn't have a dataset

Reply:
Evaluator = wrapper for a dataloader to include metrics that apply to a specific dataset